rfd/0095-application-access-okta-integration.md

```diff
@@ -0,0 +1,315 @@
+to run roughly every 2 minutes. This process wil translate Okta users, groups, and applications
+into their Teleport equivalents.
+
+```mermaid
+sequenceDiagram
+  participant OktaSvc as Okta Service
+  participant Okta as Okta API
+  participant AppSvc as Application Access Service
+
+  loop Every 2 minutes
+    OktaSvc->>Okta: Get list of applications
+    OktaSvc->>AppSvc: Synchronize applications in Teleport
+
+    OktaSvc->>Okta: Get list of groups
+
+    OktaSvc->>AppSvc: Synchronize groups in Teleport
+  end
+```
+
+#### Okta to Teleport mappings
+
+Okta groups and applications will be mapped by the background synchronization into Teleport roles and
+applications respectively.
+
+##### Groups
+
+A Teleport role will be created for each Okta group. Each group will be prepended with a configurable
+`role_prefix` which defaults to `okta/` and then suffixed with the type and name of the object, delimited by
+`/` characters.  For example, if an Okta group called `Developers` was synchronized to Teleport,
+it would be called `okta/group/Developers`.
+
+These roles use a configurable `role_template` when created which defaults to allowing users access to
+applications that belong to this group. The resulting role for `okta/group/Developers` would look like:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: okta/group/Developers
+spec:
+  allow:
+    app_labels:
+      group: "okta/group/Developers"
+```
+
+The corresponding application labels will be populated by application synchronization.
+
+##### Applications
+
+When applications are synchronized with Teleport, they will be created in application access as
+HTTP apps that use the `appLinks` from Okta as their URI. If there is more than one `appLink`
+associated with an Okta application, it will be split into multiple applications for
+each `appLink` with the unique name of each `appLink` used to disambiguate them. The `origin` field
+in the application metadata will be set to `okta`.
+
+Applications in Okta can be assigned to groups, and the groups that the application is assigned
+to will be used to create labels for the application.
+
+One complicating factor for us is that Okta allows per-application user assignment as well. As a
+result, we'll need to create a role for each application with a list of users assigned to it and allow
+it to access the application which will have a label based off of the application's unique ID
+within Okta. The label and role will look like `okta/app/123456789`. Unfortunately, Okta allows
+duplicate application names, so we'll have to use the actual unique identifier here. The role will have
+the same role template that are used by groups, but with the application specific label targeted.
+
+Using the same prefixes listed in the previous section for groups, if an application `Slack` is
+assigned to `Developers` and `Sales`, the resulting application object would look like:
+
+```yaml
+kind: app
+version: v3
+metadata:
+  name: Slack
+  origin: okta
```

Comment:
I think the label name is `teleport.dev/origin`. https://github.com/gravitational/teleport/blob/master/api/types/constants.go#L330